### ISSUE TITLE
Closed connection doesn't always stop consuming

### ISSUE THREAD

**Issue by LnL7** [sozu-proxy/lapin#283]
Seen a few times with 1.0.0-beta one a deployment when rabbitmq is restarted. With debug logging I'm still seeing `send heartbeat` logging from lapin but no heartbeats are received anymore. Inspecting the process also indicates that the connection isn't open anymore.

```
$ lsof -p 25678 | grep TCP
github_co 25678 ofborg    3u     IPv4          164411956      0t0       TCP core-0.ewr1.nix.ci:48344->core-0.ewr1.nix.ci:amqps (CLOSE_WAIT)
```

**Comment by Keruspe** [comment]
Which version of lapin did you last experience this with?

**Comment by LnL7** [comment]
Oh sorry that was supposed to be `1.0.0-beta4`.

**Comment by Keruspe** [comment]
I cannot reproduce here with latest rc. There has been lots of fixes and cleanups since beta4, please try and feel free to reopen if you manage to reproduce with a more recent version on your side.

NB: 1.0 should be out by next week

**Comment by LnL7** [comment]
To clarify, this doesn't happen consistently. Anything else I could do to help debug it this if it we see it again?

**Comment by Keruspe** [comment]
Version + logs should be enough.

We now properly cancel the heartbeat and close the reactor thread though, since a few preview releases back

**Comment by LnL7** [comment]
Encountered again under a slightly different condition, this time the broker restarting was handled correctly, but the new process never got to the point of initiating a proper connection / heartbeat.

```
May 29 02:11:16.526 DEBUG lapin::channels: send heartbeat
May 29 02:11:19.772 DEBUG lapin::channels: received heartbeat from server
May 29 02:11:21.360 ERROR lapin::channel: Connection closed on channel 0 by 0:0 => AMQPError { kind: Hard(CONNECTIONFORCED), message: ShortString("CONNECTION_FORCED - broker forced connection closure with reason \'shutdown\'") } => CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'
May 29 02:11:21.361 ERROR lapin::channels: Connection error
May 29 02:11:21.361  INFO builder: Closed the session... EOF

May 29 02:11:21.547  INFO ofborg: Logging configured
May 29 02:11:21.622 DEBUG lapin::channels: create channel with id 0
```

```
$ sudo lsof -p 22256
builder 22256 ofborg    5u     IPv4 0xeeedda39ab1f1d83       0t0       TCP 10.0.2.15:51769->core-0.ewr1.nix.ci:5671 (CLOSED)
```

**Comment by Keruspe** [comment]
Which version of lapin is that?

**Comment by cole-h** [comment]
We're running `1.0.0-rc6`.

**Comment by Keruspe** [comment]
O was expecting something like that. There was a regression in the RC that introduced connection auto close on drop, which was only fixed in the final release with https://github.com/CleverCloud/lapin/commit/c6520de9cab86d4bc799e30882ae0136e92a754b

Basically, that means that the server wasn’t ready yet. If that’s the same regression as the one I detected (like it seems to be), it’s fixed in 1.0

**Comment by LnL7** [comment]
I didn't realise 1.0 was released already, we'll update again :smile:.
